scripts/triggers-end-to-end-tests/test.js

```diff
@@ -0,0 +1,284 @@
+  this.emulators_process = null;
+}
+
+/*
+ * Check that all directions of database <-> functions <-> firestore
+ * worked.
+ */
+TriggerEndToEndTest.prototype.success = function success()
+{
+  return this.rtdb_from_firestore &&
+      this.rtdb_from_rtdb &&
+      this.firestore_from_firestore &&
+      this.firestore_from_rtdb;
+};
+
+TriggerEndToEndTest.prototype.startEmulators = function startEmulators()
+{
+  var self = this;
+  self.emulators_process = subprocess.spawn("node", [
+      PROJECT_ROOT + "/lib/bin/firebase.js",
+      "emulators:start",
+      "--project",
+      FIREBASE_PROJECT
+  ]);
+
+  self.emulators_process.stdout.on("data", function (data) {
+    process.stdout.write("[emulators stdout] " + data);
+    if (data.indexOf(RTDB_FUNCTION_LOG) > -1) {
+      self.rtdb_trigger_count++;
+    }
+    if (data.indexOf(FIRESTORE_FUNCTION_LOG) > -1) {
+      self.firestore_trigger_count++;
+    }
+  });
+
+  self.emulators_process.stderr.on("data", function (data) {
+    console.log("[emulators stderr] " + data);
+  });
+}
+
+TriggerEndToEndTest.prototype.stopEmulators = function stopEmulators()
+{
+  var self = this;
+
+  return new Promise(function (resolve, reject) {
+    self.emulators_process.on('close', function (exitCode, signal) {
+      resolve();
+    });
+
+    /*
+     * CLI process only shuts down emulators cleanly on SIGINT.
+     */
+    self.emulators_process.kill("SIGINT");
+  });
+}
+
+TriggerEndToEndTest.prototype.writeToRtdb = function writeToRtdb()
+{
+  var self = this;
+
+  return new Promise(function (resolve, reject) {
+    var url = "http://localhost:" + [
+        self.functions_emulator_port,
+        FIREBASE_PROJECT,
+        FIREBASE_PROJECT_ZONE,
+        "writeToRtdb"
+    ].join("/");
+
+    request.get(url).on('response', function(response) {
+      if (response.statusCode === 200) {
+        resolve();
+      } else {
+        reject();
+      }
+    });
+  });
+};
+
+TriggerEndToEndTest.prototype.writeToFirestore = function writeToFirestore()
+{
+  var self = this;
+
+  return new Promise(function (resolve, reject) {
+    var url = "http://localhost:" + [
+        self.functions_emulator_port,
+        FIREBASE_PROJECT,
+        FIREBASE_PROJECT_ZONE,
+        "writeToFirestore"
+    ].join("/");
+
+    request.get(url).on('response', function(response) {
+      if (response.statusCode === 200) {
+        resolve();
+      } else {
+        reject();
+      }
+    });
+  });
+};
+
+function sleep(ms)
+{
+  return new Promise(function (resolve, reject) {
+    setTimeout(resolve, ms);
+  });
+}
+
+function readConfig(done)
+{
+  fs.readFile("firebase.json", function (err, data) {
+    if (err) {
+      console.log("unable to read firebase.json " + err);
+      done(err);
+      return;
+    }
+    var config;
+    try {
+      config = JSON.parse(data);
+    } catch (err) {
+      console.log("malformed firebase.json " + err);
+    }
+    done(null, config);
+  });
+}
+
+function runTest(config)
```

Comment:
Could we use a testing library (mocha, ideally) to run this?  A few benefits:
1. Could use real `assert` statements below instead of writing your own checks.
2. Support for cleaner start/cleanup with `before()` and `after()`
3. No need for `sleep()` since they support async tests with timeouts.

If you haven't used mocha before and would find this daunting let me know, I am happy to make that change as a follow-up.

Reply:
Done, only we still do need sleeps in places where we need to wait for parallel emulator subprocesses to start up.